{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "460bab00",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "decdeadc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.ipynb_checkpoints', 'GCF_000013425.1(all_ref_files)', 'ncbi_dataset', 's-aureus-gene-list.txt', 'viewer.ipynb', 'zips-backups']\n",
      "['ncbi_dataset', 'README.md']\n",
      "['FDAARGOS_5', '29b_MRSA', '31b_MRSA', '33b', '26b_MRSA', '25b_MRSA', '27b_MRSA', 'FCFHV36', 'FDAARGOS_2', 'FDAARGOS_6']\n",
      "['assembly_data_report.jsonl', 'dataset_catalog.json', 'data_summary.tsv', 'GCF_000626615.2', 'GCF_000815045.1(29b_MRSA)', 'GCF_000815085.1', 'GCF_000815125.1', 'GCF_000815165.1', 'GCF_000815205.1', 'GCF_000815245.1', 'GCF_000969225.1', 'GCF_001018645.2', 'GCF_001018655.2']\n"
     ]
    }
   ],
   "source": [
    "'''Setup'''\n",
    "print(os.listdir())\n",
    "print(os.listdir('ncbi_dataset'))\n",
    "jsonListFile = open(\"ncbi_dataset/ncbi_dataset/data/assembly_data_report.jsonl\")\n",
    "jsonList = []\n",
    "for jsonF in jsonListFile:\n",
    "    jsonList.append(json.loads(jsonF))\n",
    "print([sample['strain'] for sample in jsonList]) \n",
    "'''\n",
    "List dir of samples\n",
    "'''\n",
    "print(os.listdir(\"ncbi_dataset/ncbi_dataset/data/\"))\n",
    "\n",
    "'''\n",
    "Use gtfparse module\n",
    "'''\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05bc1635",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Ref Data (UCSC Genome Browser)'''\n",
    "refDF = pd.read_csv(\"s-aureus-gene-list.txt\", sep=\"\\t\")\n",
    "refDF = refDF.sort_values('txStart')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "65fd5247",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2697</th>\n",
       "      <td>585</td>\n",
       "      <td>SAVP001</td>\n",
       "      <td>plasmid_VRSAp</td>\n",
       "      <td>+</td>\n",
       "      <td>314</td>\n",
       "      <td>1250</td>\n",
       "      <td>314</td>\n",
       "      <td>1250</td>\n",
       "      <td>1</td>\n",
       "      <td>314,</td>\n",
       "      <td>1250,</td>\n",
       "      <td>100001</td>\n",
       "      <td>repA</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>585</td>\n",
       "      <td>SAV0001</td>\n",
       "      <td>chr</td>\n",
       "      <td>+</td>\n",
       "      <td>516</td>\n",
       "      <td>1878</td>\n",
       "      <td>516</td>\n",
       "      <td>1878</td>\n",
       "      <td>1</td>\n",
       "      <td>516,</td>\n",
       "      <td>1878,</td>\n",
       "      <td>1</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2698</th>\n",
       "      <td>585</td>\n",
       "      <td>SAVP002</td>\n",
       "      <td>plasmid_VRSAp</td>\n",
       "      <td>-</td>\n",
       "      <td>1730</td>\n",
       "      <td>2000</td>\n",
       "      <td>1730</td>\n",
       "      <td>2000</td>\n",
       "      <td>1</td>\n",
       "      <td>1730,</td>\n",
       "      <td>2000,</td>\n",
       "      <td>100002</td>\n",
       "      <td>none</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585</td>\n",
       "      <td>SAV0002</td>\n",
       "      <td>chr</td>\n",
       "      <td>+</td>\n",
       "      <td>2155</td>\n",
       "      <td>3289</td>\n",
       "      <td>2155</td>\n",
       "      <td>3289</td>\n",
       "      <td>1</td>\n",
       "      <td>2155,</td>\n",
       "      <td>3289,</td>\n",
       "      <td>2</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>585</td>\n",
       "      <td>SAVP003</td>\n",
       "      <td>plasmid_VRSAp</td>\n",
       "      <td>+</td>\n",
       "      <td>2208</td>\n",
       "      <td>4131</td>\n",
       "      <td>2208</td>\n",
       "      <td>4131</td>\n",
       "      <td>1</td>\n",
       "      <td>2208,</td>\n",
       "      <td>4131,</td>\n",
       "      <td>100003</td>\n",
       "      <td>traA</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      #bin     name          chrom strand  txStart  txEnd  cdsStart  cdsEnd  \\\n",
       "2697   585  SAVP001  plasmid_VRSAp      +      314   1250       314    1250   \n",
       "0      585  SAV0001            chr      +      516   1878       516    1878   \n",
       "2698   585  SAVP002  plasmid_VRSAp      -     1730   2000      1730    2000   \n",
       "1      585  SAV0002            chr      +     2155   3289      2155    3289   \n",
       "2699   585  SAVP003  plasmid_VRSAp      +     2208   4131      2208    4131   \n",
       "\n",
       "      exonCount exonStarts exonEnds   score name2 cdsStartStat cdsEndStat  \\\n",
       "2697          1       314,    1250,  100001  repA         cmpl       cmpl   \n",
       "0             1       516,    1878,       1  dnaA         cmpl       cmpl   \n",
       "2698          1      1730,    2000,  100002  none         cmpl       cmpl   \n",
       "1             1      2155,    3289,       2  dnaN         cmpl       cmpl   \n",
       "2699          1      2208,    4131,  100003  traA         cmpl       cmpl   \n",
       "\n",
       "     exonFrames  \n",
       "2697         0,  \n",
       "0            0,  \n",
       "2698         0,  \n",
       "1            0,  \n",
       "2699         0,  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "116b1085",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'gene', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'internal', 'db_xref', 'transcript_biotype', 'anticodon', 'promotes', 'forms', 'ends', 'misfolded', 'exception', 'part']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'db_xref', 'exception', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'db_xref', 'exception', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'db_xref', 'exception', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'exception', 'gene_synonym', 'db_xref', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'db_xref', 'exception', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'partial', 'pseudo', 'note', 'missing', 'gene_synonym', 'db_xref', 'exception', 'internal', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'anticodon', 'transcript_biotype', 'pseudo', 'note', 'partial', 'missing', 'internal', 'gene_synonym', 'db_xref', 'exception', 'promotes', 'forms', 'ends', 'misfolded']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'gene_synonym', 'pseudo', 'note', 'db_xref', 'transcript_biotype', 'anticodon', 'partial', 'missing', 'internal', 'promotes', 'forms', 'ends', 'misfolded', 'exception', 'part', 'too']\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/nick/anaconda3/envs/pafVCF/lib/python3.9/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gbkey', 'gene_biotype', 'locus_tag', 'old_locus_tag', 'inference', 'product', 'protein_id', 'transl_table', 'exon_number', 'gene', 'gene_synonym', 'partial', 'pseudo', 'note', 'missing', 'anticodon', 'transcript_biotype', 'internal', 'db_xref', 'exception', 'promotes', 'forms', 'ends', 'misfolded', 'part']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "gtfparse doesn't read in attributes correctly for some reason (delimiter is ;)\n",
    "\"\"\"\n",
    "\n",
    "path = \"ncbi_dataset/ncbi_dataset/data/\"\n",
    "gtf = \"/genomic.gtf\"\n",
    "sample_names = [sample for sample in os.listdir(path) if \"GCF\" in sample]\n",
    "gtfs = []\n",
    "for sample_name in sample_names:\n",
    "    file_name = path + sample_name + gtf\n",
    "    curr = read_gtf(file_name)\n",
    "#     curr = curr[curr.feature == \"gene\"]\n",
    "    curr = curr.reset_index(drop=True)\n",
    "#     curr.drop(labels=\"index\", axis=0)\n",
    "    gtfs.append(curr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3dbcdc7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1        WP_000523151.1\n",
       "5        WP_000730043.1\n",
       "9        WP_001185193.1\n",
       "13       WP_000451837.1\n",
       "17       WP_001051556.1\n",
       "              ...      \n",
       "11032    WP_000920239.1\n",
       "11036    WP_000691839.1\n",
       "11040    WP_001226823.1\n",
       "11045    WP_000998767.1\n",
       "11046    WP_000998767.1\n",
       "Name: protein_id, Length: 2716, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfs[0][gtfs[0][\"feature\"] == \"CDS\"][\"protein_id\"] # Bingo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8b08f985",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17      pseudogene\n",
       "116     pseudogene\n",
       "119     pseudogene\n",
       "132     pseudogene\n",
       "151     pseudogene\n",
       "           ...    \n",
       "2626    pseudogene\n",
       "2638    pseudogene\n",
       "2710    pseudogene\n",
       "2728    pseudogene\n",
       "2758    pseudogene\n",
       "Name: gene_biotype, Length: 137, dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "!!!\n",
    "Instead of searching for gene IDs, look at protein IDs which are typically from GenBank. This might\n",
    "be a more direct way of finding good info. Also, these are probably only found in the rows\n",
    "where feature == protein\n",
    "'''\n",
    "\n",
    "gtfs[0][gtfs[0][\"gene_biotype\"] != \"protein_coding\"][\"gene_biotype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0801ba4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "parseIdcs = [ide.index(\"_\") for ide in gtfs[0][\"gene_id\"]]\n",
    "geneList = pd.Series([identity[parseIdcs[idx] + 1:] for idx, identity in enumerate(gtfs[0][\"gene_id\"])])\n",
    "gtfs[0][\"parsed_gene_id\"] = geneList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ca642808",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2790,)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneList.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "be4ee3fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>internal</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>promotes</th>\n",
       "      <th>forms</th>\n",
       "      <th>ends</th>\n",
       "      <th>misfolded</th>\n",
       "      <th>exception</th>\n",
       "      <th>part</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>296</td>\n",
       "      <td>1417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00010</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>296</td>\n",
       "      <td>1414</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00010</td>\n",
       "      <td>unassigned_transcript_1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>296</td>\n",
       "      <td>298</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00010</td>\n",
       "      <td>unassigned_transcript_1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1415</td>\n",
       "      <td>1417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00010</td>\n",
       "      <td>unassigned_transcript_1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>1528</td>\n",
       "      <td>2409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00015</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11044</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00005</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11045</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2808967</td>\n",
       "      <td>2810186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>X998_RS00005</td>\n",
       "      <td>unassigned_transcript_2789</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11046</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00005</td>\n",
       "      <td>unassigned_transcript_2789</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11047</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>2808967</td>\n",
       "      <td>2808969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00005</td>\n",
       "      <td>unassigned_transcript_2789</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11048</th>\n",
       "      <td>NZ_CP007539.2</td>\n",
       "      <td>Protein Homology</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>17</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>X998_RS00005</td>\n",
       "      <td>unassigned_transcript_2789</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11049 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqname            source      feature    start      end  score  \\\n",
       "0      NZ_CP007539.2            RefSeq         gene      296     1417    NaN   \n",
       "1      NZ_CP007539.2  Protein Homology          CDS      296     1414    NaN   \n",
       "2      NZ_CP007539.2  Protein Homology  start_codon      296      298    NaN   \n",
       "3      NZ_CP007539.2  Protein Homology   stop_codon     1415     1417    NaN   \n",
       "4      NZ_CP007539.2            RefSeq         gene     1528     2409    NaN   \n",
       "...              ...               ...          ...      ...      ...    ...   \n",
       "11044  NZ_CP007539.2            RefSeq         gene        1       19    NaN   \n",
       "11045  NZ_CP007539.2  Protein Homology          CDS  2808967  2810186    NaN   \n",
       "11046  NZ_CP007539.2  Protein Homology          CDS        1       16    NaN   \n",
       "11047  NZ_CP007539.2  Protein Homology  start_codon  2808967  2808969    NaN   \n",
       "11048  NZ_CP007539.2  Protein Homology   stop_codon       17       19    NaN   \n",
       "\n",
       "      strand frame       gene_id               transcript_id  ... internal  \\\n",
       "0          +     0  X998_RS00010                              ...            \n",
       "1          +     0  X998_RS00010     unassigned_transcript_1  ...            \n",
       "2          +     0  X998_RS00010     unassigned_transcript_1  ...            \n",
       "3          +     0  X998_RS00010     unassigned_transcript_1  ...            \n",
       "4          +     0  X998_RS00015                              ...            \n",
       "...      ...   ...           ...                         ...  ...      ...   \n",
       "11044      +     0  X998_RS00005                              ...            \n",
       "11045      +     2  X998_RS00005  unassigned_transcript_2789  ...            \n",
       "11046      +     0  X998_RS00005  unassigned_transcript_2789  ...            \n",
       "11047      +     0  X998_RS00005  unassigned_transcript_2789  ...            \n",
       "11048      +     0  X998_RS00005  unassigned_transcript_2789  ...            \n",
       "\n",
       "      db_xref transcript_biotype anticodon promotes forms ends misfolded  \\\n",
       "0                                                                          \n",
       "1                                                                          \n",
       "2                                                                          \n",
       "3                                                                          \n",
       "4                                                                          \n",
       "...       ...                ...       ...      ...   ...  ...       ...   \n",
       "11044                                                                      \n",
       "11045                                                                      \n",
       "11046                                                                      \n",
       "11047                                                                      \n",
       "11048                                                                      \n",
       "\n",
       "      exception part  \n",
       "0                     \n",
       "1                     \n",
       "2                     \n",
       "3                     \n",
       "4                     \n",
       "...         ...  ...  \n",
       "11044              2  \n",
       "11045              1  \n",
       "11046              2  \n",
       "11047                 \n",
       "11048                 \n",
       "\n",
       "[11049 rows x 35 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "287e1d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_overlapping(x1,x2,y1,y2):\n",
    "    return max(x1,y1) <= min(x2,y2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b9a9eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Function for determining the best gene match in the reference dataframe to the query dataframe.\n",
    "\n",
    "Mark query[\"likelygene\"] to ref[\"name\"]\n",
    "\n",
    "To Do:\n",
    "    - Declare matches if query and ref are on the same strand\n",
    "    - Do some bs parallelization to speed up testing time\n",
    "    - Could use binary search since this is a search on sorted arrays, but I'm lazy\n",
    "'''\n",
    "\n",
    "qtups = pd.concat([queryDF[\"start\"], queryDF[\"end\"], queryDF[\"strand\"]], axis = 1)\n",
    "rtups = pd.concat([refDF[\"txStart\"], refDF[\"txEnd\"], refDF[\"strand\"]], axis = 1)\n",
    "\n",
    "matchList = []\n",
    "likelygenes = []\n",
    "for qtup in qtups.index: # iterates over query df indices\n",
    "    potentialMatch = []\n",
    "    qS, qE, qstrand = qtups.loc[qtup]\n",
    "    for rtup in rtups.index: # # iterates over ref df indices\n",
    "        rS, rE, rstrand = rtups.loc[rtup]\n",
    "        if is_overlapping(qS, qE, rS, rE) and qstrand == rstrand:\n",
    "#             print(f\"{qS,qE} might match with {rS,rE}\")\n",
    "            potentialMatch.append([qS,qE, rS,rE, rtup])\n",
    "    if len(potentialMatch) == 0: # If there are no matches\n",
    "        continue\n",
    "    else:\n",
    "#         print( max(potentialMatch, key= lambda x: min(x[1], x[3]) - max(x[0], x[2]))) # Finds potential match with most overlap\n",
    "        likelygenepos = max(potentialMatch, key= lambda x: min(x[1], x[3]) - max(x[0], x[2]))\n",
    "        matchList.append(likelygenepos) # Creates a list containing a pair of gene coordinates which are likely matches\n",
    "        likelygenes.append(refDF[\"name\"].loc[likelygenepos[-1]] if len(potentialMatch) != 0 else \"missing\")\n",
    "#         print(likelygenes)\n",
    "#         likelygenes.append(refDF[\"name\"].loc[likelygenepos[-1]]) # Creates a list containing a likely gene for each query entry\n",
    "print(matchList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8462504",
   "metadata": {},
   "outputs": [],
   "source": [
    "matchListSave = matchList.copy()\n",
    "matchListSave"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da0b46c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def geneIntersect(query: \"dataframe\", ref: \"dataframe\") -> \"dataframe\":\n",
    "    '''\n",
    "    Function for querying gene locations to dermine annotated genes from found gene locations\n",
    "    Gonna have to go query to many times on ref in terms of dataframes\n",
    "     \n",
    "    Each query entry has to check for an intersection using formula in https://stackoverflow.com/questions/3269434/whats-the-most-efficient-way-to-test-if-two-ranges-overlap\n",
    "    for all entries in query\n",
    "    \n",
    "    Annotate query rows which correspond to a gene, otherwise make default entry something like MISSING.\n",
    "    This will help in synteny calculations.\n",
    "    \n",
    "    query: start, end\n",
    "    ref:   txstart, txend\n",
    "    '''\n",
    "    \n",
    "    qStarts, qEnds = query[\"start\"], query[\"end\"] # series\n",
    "    rStarts, rEnds = ref[\"txstart\"], ref[\"txend\"] # series\n",
    "    \n",
    "    for qS, qE in (qStarts, qEnds):\n",
    "        for rS, rE in (rStarts, rEnds):\n",
    "            if qS > rE:\n",
    "                break\n",
    "            \n",
    "    \n",
    "    \n",
    "    return query[\"start\"]\n",
    "\n",
    "    \n",
    "# def isIntersect(start: \"int\", end: \"int\", strand: \"str\", ref: \"dataframe\") -> \"bool\":\n",
    "#     if strand != ref['strand']:\n",
    "#         return False\n",
    "#     else:\n",
    "#         if ref.query(\"@ref.start <= @start and @ref.end >= @end\"): # If query is between bounds\n",
    "#             return\n",
    "#         if ref.query(\"@ref.start <= @start and @ref.end <= @end\"): # If query's end is less than upper bound\n",
    "#             return True\n",
    "#         if ref.query(\"@ref.start >= @start and @ref.end >= @end\"): # If query's start is greater than lower bound\n",
    "#             return True\n",
    "#         if ref.query(\"@ref.start >= @start and @ref.end <= @end\"): # If bounds are between query start and end\n",
    "#             return True\n",
    "#         return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "661199ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "isIntersect(516, 1878, geneQueryDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edff645c",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_overlapping(516, 1878, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d53eea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneDF[ [True if (pos[0] > 516 and pos[1] < 1878 and pos[2] == \"+\") else False for pos in geneQueryDF ] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c416b6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2[df2.feature=='gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd346639",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "1.) Grab GTF annotations from each Genome (NCBI Datasets)\n",
    "2.) Grab genes list from gene positions from GTFs\n",
    "3.) Grab list of annotated reference genes from UCSC Genome Browser (Table Browser) or EMBL\n",
    "4.) Annotate Query Dataframe with Most associated gene\n",
    "3.) Build algorithm for calculating gene SI's\n",
    "4.) Build confidence bound ...\n",
    "\n",
    "- Use table browser in UCSC genome browser\n",
    "- track: Genbank RefSeq\n",
    "- region: position\n",
    "- Get output\n",
    "- Receive a table where each entry contains features relating to the captured gene \n",
    "\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pafVCF]",
   "language": "python",
   "name": "conda-env-pafVCF-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
